### PR TITLE
8306033: Resolve multiple definition of 'throwIOException' and friends when statically linking with JDK native libraries

### PR DESCRIPTION
**Pull request by jianglizhou**
- Make functions 'static' when feasible:
  - throwByName() in src/java.security.jgss/share/native/libj2gss/NativeUtil.c.
  - throwByName(), throwIOException() and throwNullPointerException() in src/java.smartcardio/unix/native/libj2pcsc/pcsc_md.c.
  - throwByName() in src/jdk.crypto.cryptoki/share/native/libj2pkcs11/p11_util.c.
  - throwOutOfMemoryError() in src/java.smartcardio/share/native/libj2pcsc/pcsc.c.
  - Move throwDisconnectedRuntimeException() to src/jdk.crypto.cryptoki/share/native/libj2pkcs11/p11_util.c since it's only used in the file. Make it static.
  - Move throw_internal_error() to src/jdk.management/unix/native/libmanagement_ext/OperatingSystemImpl.c as it's only used in the file. Make it static.

- Rename functions by following the existing naming usages in different libraries code:
  - Rename throwOutOfMemoryError() to gssThrowOutOfMemoryError() in libj2gss.
  - Rename throwOutOfMemoryError() to p11ThrowOutOfMemoryError() in libj2pks11.
  - Rename throwNullPointerException() to p11ThrowNullPointerException() in libj2pks11.
  - Rename throwIOException() to p11ThrowIOException() in libj2pks11.
  - Rename throwPKCS11RuntimeException() to p11ThrowPKCS11RuntimeException() in libj2pks11. This function only exists in libj2pks11. The rename is done so the function naming is consistent with the other throw<exception> functions.

- Remove throw_internal_error() from src/java.management/share/native/libmanagement/management.h and src/java.management/share/native/libmanagement/management.c. It's not used.
- Remove throw_internal_error() from src/jdk.management/share/native/libmanagement_ext/management_ext.h and src/jdk.management/share/native/libmanagement_ext/management_ext.c.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306033](https://bugs.openjdk.org/browse/JDK-8306033): Resolve multiple definition of 'throwIOException' and friends when statically linking with JDK native libraries

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13497/head:pull/13497` \
`$ git checkout pull/13497`

Update a local copy of the PR: \
`$ git checkout pull/13497` \
`$ git pull https://git.openjdk.org/jdk.git pull/13497/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13497`

View PR using the GUI difftool: \
`$ git pr show -t 13497`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13497.diff">https://git.openjdk.org/jdk/pull/13497.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13497#issuecomment-1511781826)

### CODE REVIEW

**Comment by jaikiran** [comment]
Hello Jiangli, I have no experience in C language or the symbol linking that's involved in this PR, but given similar PRs that have been merged recently, is there some way these issues can be caught when code changes are being reviewed in future? Or would this require explicit run of some tool to identify these issues? I realize that you have been working on JDK-8303796 which proposes to build statically linked JDK image. Once that work is done and integrated in the JDK build, would that catch issues like these across different components of the JDK, before such changes get merged in?

**Comment by jianglizhou** [comment]
Hi Jaikiran,

Thanks for looking into this.

> is there some way these issues can be caught when code changes are being reviewed in future? Or would this require explicit run of some tool to identify these issues?

The symbol issues are reported as linker errors when statically linking the executable with JDK native libraries and libjvm. I just created a branch with the preliminary makefile changes for optionally statically linking JDK, which can be used to demonstrate/identify the issues:

  https://github.com/jianglizhou/jdk/pull/new/JDK-8303796

The branch is based on our prototype on JDK 11, but mostly reworked. It re-uses the existing JDK make infrastructure based on the feedback from Severin Gehwolf in https://bugs.openjdk.org/browse/JDK-8303796. Still need to clean up the makefile changes before making it ready for review. Early feedback/input is also welcomed.

> Once that work is done and integrated in the JDK build, would that catch issues like these across different components of the JDK, before such changes get merged in?

Yes, one option would be to include the JDK static build in pre-submit testing, which can catch any changes breaking the build. That would need to be discussed broadly with other members in OpenJDK.

**Comment by jaikiran** [comment]
Thank you for those details, Jiangli.

> 
> 
> > Once that work is done and integrated in the JDK build, would that catch issues like these across different components of the JDK, before such changes get merged in?
> 
> Yes, one option would be to include the JDK static build in pre-submit testing, which can catch any changes breaking the build. That would need to be discussed broadly with other members in OpenJDK.

Even if it doesn't get enrolled in the pre-submit testing, I think the fact that there will be a build within the JDK which can catch these issues is a good thing. It might catch the issue(s) after the integration, but I think it's still a good improvement to have these issues identified by running a specific variant of the JDK build process.

**Comment by jianglizhou** [comment]
> Even if it doesn't get enrolled in the pre-submit testing, I think the fact that there will be a build within the JDK which can catch these issues is a good thing. It might catch the issue(s) after the integration, but I think it's still a good improvement to have these issues identified by running a specific variant of the JDK build process.

Agreed, thanks.

File: src/java.security.jgss/share/native/libj2gss/GSSLibStub.c [hunk 1]
```diff
@@ -198,7 +198,7 @@ gss_channel_bindings_t newGSSCB(JNIEnv *env, jobject jcb) {
 
   cb = malloc(sizeof(struct gss_channel_bindings_struct));
   if (cb == NULL) {
-    throwOutOfMemoryError(env,NULL);
+    gssThrowOutOfMemoryError(env,NULL);
```
Comment (mcpowers) [inline src/java.security.jgss/share/native/libj2gss/GSSLibStub.c:201]: While you're fixing this, add a space between arguments, e.g. `(env,NULL) `becomes `(env, NULL)`.

File: src/java.security.jgss/share/native/libj2gss/NativeUtil.c [hunk 2]
```diff
@@ -453,15 +453,15 @@ jint getJavaErrorCode(int cNonCallingErr) {
 
 
 /* Throws a Java Exception by name */
-void throwByName(JNIEnv *env, const char *name, const char *msg) {
+static void throwByName(JNIEnv *env, const char *name, const char *msg) {
```
Comment (mcpowers) [inline src/java.security.jgss/share/native/libj2gss/NativeUtil.c:456]: Why can't you move the few lines of `throwByName()` into `gssThrowOutOfMemoryError()` and totally remove `throwByName()`?

**Review by mcpowers — COMMENTED** [review]
Update copyrights to 2023.

File: src/java.security.jgss/share/native/libj2gss/GSSLibStub.c (hunk 1 above)
Comment (jianglizhou) [reply src/java.security.jgss/share/native/libj2gss/GSSLibStub.c]: Done, thanks.

File: src/java.security.jgss/share/native/libj2gss/NativeUtil.c (hunk 2 above)
Comment (jianglizhou) [reply src/java.security.jgss/share/native/libj2gss/NativeUtil.c]: Updated as suggested, thanks.

**Comment by jianglizhou** [comment]
> Update copyrights to 2023.

Done, thanks.

**Review by AlanBateman — APPROVED** [review]
I think this looks okay.

**Comment by jianglizhou** [comment]
Thanks for the reviews.

**Pull request closed without merging**